crates/plugins/mediator-coordination/src/forward/routing.rs

```diff
@@ -0,0 +1,183 @@
+    {
+        Some(connection) => connection,
+        None => {
+            let response = (
+                StatusCode::UNAUTHORIZED,
+                MediationError::UncoordinatedSender.json(),
+            );
+            return Err(response.into_response());
+        }
+    };
+
+    // check if sender's did in mediator's keylist
+    let keylist_entries = connection.keylist.iter().find(|keys| keys == &&sender_did);
+    match keylist_entries {
+        Some(_) => {
+            // store message attachement with associated recipient did
+            let message = payload.clone().attachments.expect("expect attachements");
+            let receivering_dids = next;
+            for did in receivering_dids {
+                let messages = Messages {
+                    id: None,
+                    message: message.clone(),
+                    recipient_did: did,
+                };
+                message_repository
+                    .store(messages)
+                    .await
+                    .map_err(|_| MediationError::PersisenceError)
+                    .unwrap();
+            }
+        }
+        None => {
+            let response = (
+                StatusCode::UNAUTHORIZED,
+                MediationError::UncoordinatedSender.json(),
+            );
+            return Err(response.into_response());
+        }
+    }
+
+    Ok(None)
+}
+
+#[cfg(test)]
+mod test {
+    use std::sync::Arc;
+
+    use crate::{
+        constant::MEDIATE_FORWARD_2_0,
+        forward::ledger::{ALICE_DID_DOC, ALICE_SECRETS, MEDIATOR_DID_DOC},
+        repository::stateful::coord::tests::{
+            MockConnectionRepository, MockMessagesRepository, MockSecretsRepository,
+        },
+        util::{self, MockFileSystem},
+        web::AppStateRepository,
+    };
+
+    use super::*;
+
+    use didcomm::{
+        did::resolvers::ExampleDIDResolver, secrets::resolvers::ExampleSecretsResolver, Attachment,
+        AttachmentData, JsonAttachmentData, Message, PackEncryptedOptions, UnpackOptions,
+    };
+    use uuid::Uuid;
+    pub fn setup() -> Arc<AppState> {
+        let public_domain = String::from("http://alice-mediator.com");
+
+        let mut mock_fs = MockFileSystem;
+        let diddoc = util::read_diddoc(&mock_fs, "").unwrap();
+        let keystore = util::read_keystore(&mut mock_fs, "").unwrap();
+
+        let repository = AppStateRepository {
+            connection_repository: Arc::new(MockConnectionRepository::from(vec![])),
+            secret_repository: Arc::new(MockSecretsRepository::from(vec![])),
+            message_repository: Arc::new(MockMessagesRepository::from(vec![])),
+        };
+
+        let state = Arc::new(AppState::from(
+            public_domain,
+            diddoc,
+            keystore,
+            Some(repository),
+        ));
+
+        state
+    }
+
+    #[tokio::test]
+    async fn test_mediator_forward_process() {
+        // simulate sender forwarding process
+        let did_resolver =
```

Comment:
- Make you test more realistic by reusing existing APIs of the DIDComm library.
- You're using the same DID for the sender and the recipient, which is awkward.
- Create aliases (variables or functions) to make the story line more readable.
- Normally unpacking should be done by the app, not in the test. But we can fix that later.

Check this alternative:

```rust
#[cfg(test)]
mod test {
    use std::sync::Arc;

    use crate::{
        didcomm::bridge::LocalSecretsResolver,
        repository::stateful::coord::tests::{
            MockConnectionRepository, MockMessagesRepository, MockSecretsRepository,
        },
        util::{self, MockFileSystem},
        web::AppStateRepository,
    };

    use super::*;

    use did_utils::jwk::Jwk;
    use didcomm::{
        algorithms::AnonCryptAlg, protocols::routing::wrap_in_forward, secrets::SecretsResolver, Message, PackEncryptedOptions, UnpackOptions
    };
    use uuid::Uuid;

    pub fn setup() -> Arc<AppState> {
        let public_domain = String::from("http://alice-mediator.com");

        let mut mock_fs = MockFileSystem;
        let diddoc = util::read_diddoc(&mock_fs, "").unwrap();
        let keystore = util::read_keystore(&mut mock_fs, "").unwrap();

        let repository = AppStateRepository {
            connection_repository: Arc::new(MockConnectionRepository::from(vec![])),
            secret_repository: Arc::new(MockSecretsRepository::from(vec![])),
            message_repository: Arc::new(MockMessagesRepository::from(vec![])),
        };

        let state = Arc::new(AppState::from(
            public_domain,
            diddoc,
            keystore,
            Some(repository),
        ));

        state
    }

    #[tokio::test]
    async fn test_mediator_forward_process() {
        let state = &setup();

        /* Simulate sender forwarding process */

        let msg = Message::build(
            Uuid::new_v4().to_string(),
            "example/v1".to_owned(),
            json!("Hey there! Just wanted to remind you to step outside for a bit. A little fresh air can do wonders for your mood."),
        )
        .to(_recipient_did())
        .from(_sender_did())
        .finalize();

        let (msg, _) = msg
            .pack_encrypted(
                &_recipient_did(),
                Some(&_sender_did()),
                None,
                &state.did_resolver,
                &_sender_secrets_resolver(),
                &PackEncryptedOptions::default(),
            )
            .await
            .expect("Unable pack_encrypted");

        let msg = wrap_in_forward(
            &msg,
            None,
            &_recipient_did(),
            &vec![_mediator_did(state)],
            &AnonCryptAlg::default(),
            &state.did_resolver,
        )
        .await
        .expect("Unable wrap_in_forward");

        // Mediator in action
        let (payload, _) = Message::unpack(
            &msg,
            &state.did_resolver,
            &state.secrets_resolver,
            &UnpackOptions::default(),
        )
        .await
        .unwrap();

        assert!(mediator_forward_process(state, payload).await.is_ok());
    }

    pub fn _sender_did() -> String {
        "did:key:z6MkwKfDFAK49Lb9D6HchFiCXdcurRUSFrbnwDBk5qFZeHA3".to_string()
    }

    pub fn _mediator_did(state: &AppState) -> String {
        state.diddoc.id.clone()
    }

    pub fn _recipient_did() -> String {
        "did:key:z6MkfyTREjTxQ8hUwSwBPeDHf3uPL3qCjSSuNPwsyMpWUGH7".to_string()
    }

    pub fn _sender_secrets_resolver() -> impl SecretsResolver {
        let secret_id = _sender_did() + "#z6LSiZbfm5L5zR3mrqpHyL7T2b2x3afUMpmGnMrEQznAz5F3";
        let secret: Jwk = serde_json::from_str(
            r#"{
                "kty": "OKP",
                "crv": "X25519",
                "x": "ZlJzHqy2dLrDQNlV15O3zDOIXpWVQnq6VtiVZ78O0hY",
                "d": "8OK7-1IVMdcM86PZzYKsbIi3kCJ-RxI8XFKe9JEcF2Y"
            }"#,
        )
        .unwrap();

        LocalSecretsResolver::new(&secret_id, &secret)
    }
}
```